src/back/glsl.rs

```diff
@@ -1632,3 +1663,191 @@ fn write_format_glsl(format: StorageFormat) -> &'static str {
         StorageFormat::Rgba32Float => "rgba32f",
     }
 }
+
+fn collect_texture_mapping(
+    module: &Module,
+    functions: &FastHashMap<Handle<Function>, String>,
+) -> Result<Vec<TextureMapping>, Error> {
+    fn collect_texture_mapping_expr(
+        func: &Function,
+        expr: Handle<Expression>,
+        mappings: &mut Vec<TextureMapping>,
+    ) -> Result<(), Error> {
+        match func.expressions[expr] {
+            Expression::Access { base, index } => {
+                collect_texture_mapping_expr(func, base, mappings)?;
+                collect_texture_mapping_expr(func, index, mappings)?;
+            }
+            Expression::AccessIndex { base, .. } => {
+                collect_texture_mapping_expr(func, base, mappings)?
+            }
+            Expression::Compose { ref components, .. } => {
+                for comp in components {
+                    collect_texture_mapping_expr(func, *comp, mappings)?
+                }
+            }
+            Expression::Load { pointer } => collect_texture_mapping_expr(func, pointer, mappings)?,
+            Expression::ImageSample {
+                image,
+                sampler,
+                coordinate,
+                level,
+                depth_ref,
+            } => {
+                let tex_handle = match func.expressions[image] {
+                    Expression::GlobalVariable(global) => global,
+                    // Temp
+                    Expression::Load { pointer } => match func.expressions[pointer] {
+                        Expression::GlobalVariable(global) => global,
+                        _ => unreachable!(),
+                    },
+                    _ => unreachable!(),
+                };
+
+                let sampler_handle = match func.expressions[sampler] {
+                    Expression::GlobalVariable(global) => global,
+                    // Temp
+                    Expression::Load { pointer } => match func.expressions[pointer] {
+                        Expression::GlobalVariable(global) => global,
+                        _ => unreachable!(),
+                    },
+                    _ => unreachable!(),
+                };
+
+                collect_texture_mapping_expr(func, coordinate, mappings)?;
+                match level {
+                    crate::SampleLevel::Exact(expr) | crate::SampleLevel::Bias(expr) => {
+                        collect_texture_mapping_expr(func, expr, mappings)?
+                    }
+                    crate::SampleLevel::Auto => {}
+                }
+                if let Some(expr) = depth_ref {
+                    collect_texture_mapping_expr(func, expr, mappings)?;
+                }
+
+                let mapping = mappings.iter().find(|map| map.texture == tex_handle);
```

Comment:
If I understand correctly, this is an internal map we are collecting. It looks like we are often searching it by the texture handle. Would it make things easier if this internal map was `FastHashMap<Handle<GlobalVar>, Handle<GlobalVar>>`? (addressable by texture handles)